angular/src/app/users/change-password/change-password.component.ts

```diff
@@ -0,0 +1,89 @@
+import { Component, OnInit, Injector, ViewChild, ElementRef, AfterViewInit } from '@angular/core';
+import { appModuleAnimation } from '@shared/animations/routerTransition';
+import { AppComponentBase } from '@shared/app-component-base';
+import { ChangePasswordDto, UserServiceProxy } from '@shared/service-proxies/service-proxies';
+import { Router } from '@angular/router';
+import { FormGroup, FormControl, Validators, AbstractControl, ValidationErrors, FormGroupDirective, NgForm } from '@angular/forms';
+import { finalize } from 'rxjs/operators';
+import { ErrorStateMatcher } from '@angular/material';
+
+export class FormGroupErrorStateMatcher implements ErrorStateMatcher {
+    constructor(private formGroup: FormGroup) { }
+
+    isErrorState(control: FormControl | null, form: FormGroupDirective | NgForm | null): boolean {
+        return control && control.dirty && control.touched && this.formGroup && this.formGroup.errors && this.formGroup.errors.areEqual;
+    }
+}
+
+@Component({
+    animations: [appModuleAnimation()],
+    templateUrl: './change-password.component.html'
+})
+export class ChangePasswordComponent extends AppComponentBase implements OnInit {
+    @ViewChild('body') bodyElement: ElementRef;
+    public parentFormGroup: FormGroup;
+    public passwordsFormGroup: FormGroup;
+    public isLoading: boolean;
+    public equalMatcher: FormGroupErrorStateMatcher;
+
+    private static areEqual(c: AbstractControl): ValidationErrors | null {
+        const keys: string[] = Object.keys(c.value);
+        for (const i in keys) {
+            if (i !== '0' && c.value[keys[+i - 1]] !== c.value[keys[i]]) {
+                return { areEqual: true };
+            }
+        }
+    }
+
+    public constructor(
+        injector: Injector,
+        private userServiceProxy: UserServiceProxy,
+        private router: Router
+    ) {
+        super(injector);
+    }
+
+    ngOnInit() {
+        this.isLoading = true;
+
+        this.passwordsFormGroup = new FormGroup({
+            'newPassword': new FormControl('', [
+                Validators.required,
+                Validators.pattern('(?=^.{8,}$)(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?!.*\s)[0-9a-zA-Z!@#$%^&*()]*$') ]),
```

Comment:
I hate that this regex pattern is duplicated on the front-end and the back-end, but I kinda thought it was worth it for ability to have front-end and back-end validation ... dunno, game for other ideas.